src/Auxiliary/Aux_Check_Conservation.cpp

```diff
@@ -401,131 +414,136 @@ void Aux_Check_Conservation( const char *comment )
 #     if ( defined MASSIVE_PARTICLES  &&  MODEL != PAR_ONLY )
-      const double Mass_All    = Fluid_AllRank[       0] + Mass_Par;
-      const double MomX_All    = Fluid_AllRank[       1] + MomX_Par;
-      const double MomY_All    = Fluid_AllRank[       2] + MomY_Par;
-      const double MomZ_All    = Fluid_AllRank[       3] + MomZ_Par;
-      const double AngMomX_All = Fluid_AllRank[       4] + AngMomX_Par;
-      const double AngMomY_All = Fluid_AllRank[       5] + AngMomY_Par;
-      const double AngMomZ_All = Fluid_AllRank[       6] + AngMomZ_Par;
-      const double Etot_All    = Fluid_AllRank[idx_etot] + Etot_Par;    // for HYDRO/ELBDM, total energy is stored in the last element
-
-      const double CoMX_All    = ( Fluid_AllRank[0]*CoM_Gas[0] + Mass_Par*CoMX_Par )/Mass_All;
-      const double CoMY_All    = ( Fluid_AllRank[0]*CoM_Gas[1] + Mass_Par*CoMY_Par )/Mass_All;
-      const double CoMZ_All    = ( Fluid_AllRank[0]*CoM_Gas[2] + Mass_Par*CoMZ_Par )/Mass_All;
-#     endif // if ( defined MASSIVE_PARTICLES  &&  MODEL != PAR_ONLY )
+      for (int v=0; v<7; v++)   All_AllRank[v] = Fluid_AllRank[v] + Par_AllRank[v];
+      All_AllRank[idx_etot_all] = Fluid_AllRank[idx_etot_flu] + Par_AllRank[idx_etot_par]; // for HYDRO/ELBDM, total energy is stored in the last element
 
-//    note that a variable length array cannot have static storage duration
-      static double Time_Ref;
-      static double Fluid_Ref[NVar_Max];
-      static double CoM_Gas_Ref[3];
-#     ifdef MASSIVE_PARTICLES
-      static double Mass_Par_Ref, CoMX_Par_Ref, CoMY_Par_Ref, CoMZ_Par_Ref;
-      static double MomX_Par_Ref, MomY_Par_Ref, MomZ_Par_Ref;
-      static double AngMomX_Par_Ref, AngMomY_Par_Ref, AngMomZ_Par_Ref, Ekin_Par_Ref, Epot_Par_Ref, Etot_Par_Ref;
-#     if ( MODEL != PAR_ONLY )
-      static double Mass_All_Ref, CoMX_All_Ref, CoMY_All_Ref, CoMZ_All_Ref;
-      static double MomX_All_Ref, MomY_All_Ref, MomZ_All_Ref;
-      static double AngMomX_All_Ref, AngMomY_All_Ref, AngMomZ_All_Ref, Etot_All_Ref;
-#     endif
-#     endif // #ifdef PARTICLE
-      double AbsErr[NVar], RelErr[NVar];
+      for (int d=0; d<3; d++)
+         CoM_All[d] = ( Fluid_AllRank[0]*CoM_Flu[d] + Par_AllRank[0]*CoM_Par[d] )/All_AllRank[0];
+#     endif // if ( defined MASSIVE_PARTICLES  &&  MODEL != PAR_ONLY )
 
-      if ( FirstTime )
+//    record the reference values if not initialized, e.g., first time or restart from an HDF5 snapshot with version < 2502
+      if ( ! ConRefInitialized )
       {
-//       record the reference values
-         Time_Ref = Time[0];
-         for (int v=0; v<NVar; v++)    Fluid_Ref[v]   = Fluid_AllRank[v];
-         for (int d=0; d<3; d++)       CoM_Gas_Ref[d] = CoM_Gas[d];
+         ConRef[0] = Time[0];
+         for (int v=0; v<NVar_Flu; v++)   ConRef[idx_offset_flu    +v] = Fluid_AllRank[v];
+         for (int d=0; d<3; d++)          ConRef[idx_offset_flu_com+d] = CoM_Flu[d];
```

Comment:
Can we add several sanity checks below to detect overflow? They could be similar to `Aux_Error( ERROR_INFO, "exceed NFIELD_STORED_MAX (%d) !!\n", NFIELD_STORED_MAX );` in `Output_DumpData_Total_HDF5.cpp`